File: app/services/job_url.py
"""Normalize job posting URLs for deduplication."""

from __future__ import annotations

import re
from urllib.parse import parse_qsl, urlparse, urlunparse

LINKEDIN_JOB_PATH = re.compile(r"/jobs/view/(\d+)")
TRACKING_PARAMS = frozenset(
    {
        "utm_source",
        "utm_medium",
        "utm_campaign",
        "utm_term",
        "utm_content",
        "ref",
        "refid",
        "trackingid",
        "src",
        "gh_src",
        "lever-source",
    }
)


def _linkedin_job_id(url: str) -> str | None:
    parsed = urlparse(url)
    path_match = LINKEDIN_JOB_PATH.search(parsed.path)
    if path_match:
        return path_match.group(1)
    for key, value in parse_qsl(parsed.query, keep_blank_values=True):
        if key == "currentJobId" and value:
            return value
    return None


def normalize_job_url(url: str | None) -> str | None:
    """Return a canonical URL for dedupe lookups, or None when URL is absent."""
    if url is None:
        return None
    raw = url.strip()
    if not raw:
        return None

    linkedin_id = _linkedin_job_id(raw)
    if linkedin_id:
        return f"https://www.linkedin.com/jobs/view/{linkedin_id}/"

    parsed = urlparse(raw)
    if not parsed.scheme or not parsed.netloc:
        return raw

    host = parsed.hostname.lower() if parsed.hostname else ""
    path = parsed.path.rstrip("/") or "/"
    query_pairs = [
        (key, value)
        for key, value in parse_qsl(parsed.query, keep_blank_values=True)
        if key.lower() not in TRACKING_PARAMS
    ]
    query = "&".join(f"{key}={value}" for key, value in query_pairs) if query_pairs else ""

    return urlunparse((parsed.scheme.lower(), host, path, "", query, ""))

File: frontend/src/lib/jobUrl.ts
/** Mirror backend `normalize_job_url` for client-side duplicate checks. */

const LINKEDIN_JOB_PATH = /\/jobs\/view\/(\d+)/;
const TRACKING_PARAMS = new Set([
  "utm_source",
  "utm_medium",
  "utm_campaign",
  "utm_term",
  "utm_content",
  "ref",
  "refid",
  "trackingid",
  "src",
  "gh_src",
  "lever-source",
]);

function linkedInJobId(url: string): string | null {
  try {
    const parsed = new URL(url);
    const pathMatch = parsed.pathname.match(LINKEDIN_JOB_PATH);
    if (pathMatch?.[1]) {
      return pathMatch[1];
    }
    return parsed.searchParams.get("currentJobId");
  } catch {
    return null;
  }
}

export function normalizeJobUrl(url: string | null | undefined): string | null {
  if (!url) {
    return null;
  }
  const raw = url.trim();
  if (!raw) {
    return null;
  }

  const linkedinId = linkedInJobId(raw);
  if (linkedinId) {
    return `https://www.linkedin.com/jobs/view/${linkedinId}/`;
  }

  try {
    const parsed = new URL(raw);
    if (!parsed.hostname) {
      return raw;
    }
    const host = parsed.hostname.toLowerCase();
    const path = parsed.pathname.replace(/\/$/, "") || "/";
    const params = new URLSearchParams();
    parsed.searchParams.forEach((value, key) => {
      if (!TRACKING_PARAMS.has(key.toLowerCase())) {
        params.set(key, value);
      }
    });
    const query = params.toString();
    return `${parsed.protocol}//${host}${path}${query ? `?${query}` : ""}`;
  } catch {
    return raw;
  }
}

export interface DuplicateJobInfo {
  id: string;
  title: string;
  company: string;
}

export function duplicateJobFromApiDetail(detail: unknown): DuplicateJobInfo | null {
  if (!detail || typeof detail !== "object") {
    return null;
  }
  const record = detail as Record<string, unknown>;
  if (typeof record.job_id !== "string") {
    return null;
  }
  return {
    id: record.job_id,
    title: typeof record.title === "string" ? record.title : "Existing job",
    company: typeof record.company === "string" ? record.company : "",
  };
}

export function duplicateJobFromErrorMessage(message: string): DuplicateJobInfo | null {
  const trimmed = message.trim();
  if (!trimmed.startsWith("{")) {
    return null;
  }
  try {
    return duplicateJobFromApiDetail(JSON.parse(trimmed));
  } catch {
    return null;
  }
}
